src/models/partial_tx.ts

```diff
@@ -0,0 +1,440 @@
+/**
+ * Copyright (c) Hathor Labs and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+import Input from './input';
+import Output from './output';
+import Address from './address';
+import Transaction from './transaction';
+import Network from './network';
+
+import P2PKH from './p2pkh';
+import P2SH from './p2sh';
+import ScriptData from './script_data';
+
+import transaction from '../transaction';
+import helpers from '../utils/helpers';
+import { IndexOOBError, UnsupportedScriptError } from '../errors';
+
+import txApi from '../api/txApi';
+import { TOKEN_AUTHORITY_MASK, TOKEN_INDEX_MASK, HATHOR_TOKEN_CONFIG } from '../constants';
+
+type InputData = {
+  tx_id: string,
+  index: number,
+  address?: string,
+  data: Buffer,
+};
+
+type OutputData = {
+  type: string,
+  value: number,
+  tokenData: number,
+  address?: string,
+  data?: string,
+  timelock?: number,
+};
+
+type TxData = {
+  inputs: InputData[],
+  outputs: OutputData[],
+  tokens: string[],
+  weight?: number,
+  nonce?: number,
+  version?: number,
+  timestamp?: number,
+};
+
+export class ProposalInput extends Input {
+  token: string;
+  value: number;
+  address: string;
+
+  constructor(hash: string, index: number, token: string, value: number, address: string) {
+    super(hash, index);
+    this.token = token;
+    this.value = value;
+    this.address = address;
+  }
+
+  /**
+   * Return an object with the relevant input data
+   *
+   * @return {InputData}
+   * @memberof ProposalInput
+   * @inner
+   */
+  toData() {
+    return {
+      tx_id: this.hash,
+      index: this.index,
+      data: Buffer.from([]),
```

Comment:
The `utils.helpers.createTxFromData` method assumes the data exists and should be added.
This would get an error (I don't recall where exactly) for `undefined` data.

This way we have an input data of length 0 which works with the helper and other methods and models.
And we can test `input.data.length === 0` to check if we have an input data.

Reply:
The method `createTxFromData` does not expect the `input` to have `data`. Maybe you are using this transaction to do something else rather than just `createTxFromData`.

It's good to understand this flow, maybe we should fix the other place in the lib to support this, and not just create an empty buffer

Reply:
This [line](https://github.com/HathorNetwork/hathor-wallet-lib/blob/master/src/transaction.js#L597) throws the error `TypeError: Cannot read properties of undefined (reading 'length')` since data is `undefined`.

Should I change to add zero when data is undefined?

Reply:
Now I get it. This method was supposed to be used when the tx is already completed. We can keep your code, just add some comments about it.